### ISSUE TITLE
Services: Let's Encrypt: Settings cannot Apply

### ISSUE THREAD

**Issue by baqwas** [opnsense/plugins#1948]
**Describe the bug**
A clear and concise description of what the bug is, including last known working version (if any).

*Tip*: to validate your setup was working with the previous version, use opnsense-revert (https://docs.opnsense.org/manual/opnsense_tools.html#opnsense-revert)

**To Reproduce**
1. Go to **Services: Let's Encrypt: Settings**
2. Click on **Enable Plugin**
3. Click on **Auto Renewal**
3. In _Let's Encrypt Environment_, select **Staging Environment**
4. Click button **Apply**

**Expected behavior**
A Let's Encrypt certificate to be issued in due course.

**Screenshots**
The Apply button stays dark until any portion of the pane is clicked whereupon the button reverts to original lighter color. No other visible change anywhere.

**Relevant log files**
Message on log file page:
_File /var/log/acme.sh.log doesn't exist._

**Additional context**
Full help button is enabled.

**Environment**
OPNsense 20.7-amd64
FreeBSD 12.1-RELEASE-p7-HBSD
OpenSSL 1.1.1g 21 Apr 2020

**Comment by fraenki** [comment]
> Expected behavior
> A Let's Encrypt certificate to be issued in due course.

I'm not sure that I understood this bug report correctly.... you expect that these steps would issue a Let's Encrypt certificate?

**Comment by baqwas** [comment]
Hello @fraenki,

I logged this issue because after the upgrade to 20.7, I encountered the following message while attempting to browse the log file:
_File /var/log/acme.sh.log doesn't exist._

Previously, I used to get consistently _code invalid_ in the default log file. While I don't know what this earlier message meant since the acme client tested OK but nevertheless I assumed that the issue was an operator/user error after reporting it at the forum. At that time Fabian asked me to check the logs where I saw the _code invalid_ message for the Let's Encrypt request.

I know that Let's Encrypt is fussy about port 80 being open and I can confirm that port 80 is open. I am very inexperienced in these matters. All I need is for the log file to work so that I can keep on trying to secure a Let's Encrypt cert for the OPNsense server. Thanks for looking into this report.

Kind regards.

**Comment by fraenki** [comment]
The file `/var/log/acme.sh.log` is only created when acme.sh tried so register an account or issue a certificate.
The basic workflow of the Let's Encrypt plugin is as follows:

1. Enable the plugin in `Services: Let's Encrypt: Settings` (as you presumably did)
1. Setup an account in `Services: Let's Encrypt: Accounts`
1. Configure a validation method in `Services: Let's Encrypt: Validation Methods` (DNS-01 should be preferred over HTTP-01)
1. Add a certificate in `Services: Let's Encrypt: Certificates`
1. Finally click on the `Issues/renew certificates now` button

Now you should see at least _something_ in `Services: Let's Encrypt: Log File`.

**Comment by baqwas** [comment]
Thx again @fraenki.

There is a log file now! Appreciate the detailed instructions. I'll close the issue but please allow me to post the log just for the record in case there is further insight on where I went wrong:

```
Enabled Common Name Multi-Domain (SAN) Description Issue/Renewal Date Last Acme Status Last Acme Run Commands
	opnsense.parkcircus.org	opensense.parkcircus.org	Protectli box server certificate	pending	validation failed	8/2/2020, 9:50:08 AM	
	
Date 	Message
[Sun Aug 2 09:50:07 CDT 2020] 	_CURL='curl -L --silent --dump-header /var/etc/acme-client/home/http.header --trace-ascii /tmp/tmp.P2MKLEtP -g '
[Sun Aug 2 09:50:06 CDT 2020] 	timeout=
[Sun Aug 2 09:50:06 CDT 2020] 	url='https://acme-staging-v02.api.letsencrypt.org/directory'
[Sun Aug 2 09:50:06 CDT 2020] 	GET
[Sun Aug 2 09:50:06 CDT 2020] 	_init api for server: https://acme-staging-v02.api.letsencrypt.org/directory
[Sun Aug 2 09:50:06 CDT 2020] 	Using ACME_DIRECTORY: https://acme-staging-v02.api.letsencrypt.org/directory
[Sun Aug 2 09:50:06 CDT 2020] 	DOMAIN_PATH='/var/etc/acme-client/home/opnsense.parkcircus.org'
[Sun Aug 2 09:50:06 CDT 2020] 	ACME_DIRECTORY='https://acme-staging-v02.api.letsencrypt.org/directory'
[Sun Aug 2 09:50:06 CDT 2020] 	Using stage ACME_DIRECTORY: https://acme-staging-v02.api.letsencrypt.org/directory
```
Appreciate the step-by-step instructions and fast responses. Thanks.

Kind regards.

**Comment by fraenki** [comment]
Your certificate could not be issued ("validation failed"). We would need to know your Let's Encrypt "validation method" configuration details and of course the _whole_ acme.sh.log to get an idea what went wrong.

**Comment by baqwas** [comment]
Hello @fraenki,

I've prepared a summary document (_attached_) for your review, if it helps. Please allow me to rationalize two things here although it may not be germane to the issue.

I couldn't get Let's Encrypt working initially. Then I read about using intermediate certs and I thought I should give it a try. So currently, I do have an intermediate ca. I don't know enough about this other than to let you know.

Also, I have a Synology NAS server that has an auto-renewal Let's Encrypt cert. I couldn't issue the cert myself. :) I contacted Synology tech support and provided them with screen shots of my steps. The tech support was granted temporary elevated priviledges (_per_ their support protocol) and he successfully issued the cert. When I asked what he had done, his only remark was that he had done exactly what I had tried. This cert auto renews every 3 months to this day.

Your guidance and suggestions are always welcome. Thanks.

Kind regards.
[Let's Encrypt settings.docx](https://github.com/opnsense/plugins/files/5016816/Let.s.Encrypt.settings.docx)

**Comment by DumboJet** [comment]
I am getting the same error when trying to see the  Let's Encrypt plugin logs too.
I restored my configuration into a new installation of OpnSense 20.7.7.
I saw that the Let's Encrypt plugin did not work, so I reset the client, tried again, and still nothing.
When I try to view the logs, I get the 'File /var/log/acme.sh.log doesn't exist.' message.
So, I see not easy way to find out what is wrong here.
The log file does not exist indeed (I SSH-ed into the box).
Not sure if I should reinstall the plugin or something...

**Comment by baqwas** [comment]
@DumboJet,

Thanks for the observation. I still have not been able to make Let's Encrypt work in my OPNsense (v2.7.8) installation. I am not quite sure where I went wrong because the acme clients work fine with my Ubuntu server & desktop installations on RPi hardware.

Kind regards.

**Comment by fraenki** [comment]
>  When I try to view the logs, I get the 'File /var/log/acme.sh.log doesn't exist.' message.

@DumboJet 
This log file is only created when acme.sh tries so register an account or issue a certificate. If it does not exist, then no attempt to issue a certificate was made.

Assuming that the Let's Encrypt is properly configured, there's a way to get more detailed log output:

First go to `Services: Let's Encrypt: Settings` and change Log Level to "debug". Then go to `Services: Let's Encrypt: Certificates` and click the button to forcefully reissue the certificate (it's the 3rd button).

Afterwards go to `System: Log Files: General` and search for `AcmeClient`, it should contain more detailed log messages.

> I still have not been able to make Let's Encrypt work

@baqwas 
Your initial configuration was set to HTTP-01 with "IP Auto-Discovery" enabled and interface set to "LAN". There are several requirements for this to work, some of which are mentioned in the inline help for these options:

> The FQDN's used in your certificate must currently point to an official IP address. [...]
> NOTE:This will ONLY work if the official IP addresses are LOCALLY configured on your OPNsense firewall.

In other words: if you try to issue a certificate for test.example.com, then this hostname must resolve to an IP address that is configured on your OPNsense firewall (or if you have another router than a port forward to your OPNsense firewall is required).

I doubt that this is a bug, it's more likely that it's a configuration error. I recommend the [official forums](https://forum.opnsense.org/) to search for similar reports or discuss this in-depth.

**Comment by DumboJet** [comment]
Thanks @fraenki !
I have fixed my problem.

That `Services: Let's Encrypt: Log File` section is really misleading...
My certificate is updated now and it still shows no logs.
I will have to remember that the logs go elsewhere...

**Comment by fraenki** [comment]
>  My certificate is updated now and it still shows no logs.

That should not happen and I've never seen this before. Could you check on the OPNsense commandline that the log file exists in `/var/log/acme.sh.log` or if it uses a different filename by searching in `/var/log/`?

**Comment by DumboJet** [comment]
Nope!
The file does not exist and no similar file seems present:
![image](https://user-images.githubusercontent.com/8492662/105406563-42487900-5c35-11eb-94a2-c0040e092225.png)
The service is reported to be up in the web management interface, although I am not sure what its name is, in order to check in the process list:
![image](https://user-images.githubusercontent.com/8492662/105406836-a2d7b600-5c35-11eb-86b8-3b27b77250bd.png)

But while installing OpnSense I had these issues:
https://forum.opnsense.org/index.php?topic=20957.msg97736#msg97736
So, I don't know what are the chances something is not installed correctly...

**Comment by fraenki** [comment]
>  So, I don't know what are the chances something is not installed correctly...

Probably, but I can't tell for sure.

FYI, if you have `/var RAM disk` enabled in `System: Settings: Miscellaneous`, then acme.sh.log is lost on reboot.

**Comment by DumboJet** [comment]
I do have RAM disks enabled, but I think that after the certificate update it only restarts HAProxy, so most likely I should have seen something in that log at some point.
But anyway... I have now upgraded to 20.7.8 and I will see if anything changes after many updates being applied to it.
If there was an issue at the installation, I have some hopes that it might be resolved like that automagically...

**Comment by baqwas** [comment]
Hello @fraenki,

Thanks for taking the time to update me. Sorry for the belated response. I did try the OPNsense forum (General Discussion as well Production) but did not receive any resolvable solutions. That is why as a measure of last effort I resorted to using this site (unfortunately it is a bad habit of mine to use such channels inappropriately because I don't know where else to turn).

Sorry again for taking up bandwidth here but please allow to record the changes I performed after your initial feedback:

- The Let's Encrypt CA server is unable to communicate with my OPNsense instance (which is the Edge device in my farm).
- I have no such issue with a Synology server that has been renewing LE certs every 90 days per schedule (unfortunately it does not have the schedule finesse that I like in the OPNsense instance).
- I applied a _temporary_ rule to route all 80/443 traffic from WAN to OPNsense instance (destination using LAN address). Confirmed that WAN access to ports 80/443 is open.
- Still no luck. (Obviously, my knowledge of OPNsense is at the novice level). I will keep on trying intermittently because the solution requires some improvement in my knowledge of OPNsense services.

Again, appreciate your feedback, @fraenki, and this reply is simply a comment and not a request to pursue it further here. The one bell/whistle about the Firewall Rule that appeals to me (LE cert renewal) is the schedule. This feature will help me to work with LE certs on multiple servers in an unattended mode, I presume. Thanks for your time.

Kind regards.
